File: adaptive_train.py
from Utils import readAccelerationMatrix,getMagnitude,shift_central,readAcc,readAccGyro
from time_alignment import align
from matplotlib import pyplot as plt
import csv
from RMSD import calcR,Cartesian2Spherical
import numpy as np
import os

def adaptive_train(master_raw,slave_raw):

    master=shift_central(master_raw)
    slave=shift_central(slave_raw)

    m_magnitude=[getMagnitude(master[i]) for i in range(len(master))]
    s_magnitude=[getMagnitude(slave[i]) for i in range(len(slave))]

    master=np.array(master).transpose()
    slave=np.array(slave).transpose()


    master_index=0
    slave_index=0
    k=2
    interval=50
    train_size=interval
    master_len=len(m_magnitude)
    slave_len=len(s_magnitude)
    aligned_master=[]
    aligned_slave=[]

    translated_master=[]
    translated_slave=[]
    count=0

    first=True
    while True:
        if first:
            train_size=100
            first=False
        else:
            train_size=20

        master_end=master_index+train_size
        slave_end=slave_index+train_size

        if master_end>=master_len or slave_end>=slave_len:
            l=min(max(master_end , master_len)-min(master_end , master_len),max(slave_end , slave_len)-min(slave_end , slave_len))
            aligned_master.extend(m_magnitude[master_index:master_index+l])
            aligned_slave.extend(s_magnitude[slave_index:slave_index+l])
            break
        count += 1
        m_start,s_start=align(m_magnitude[master_index:master_end],s_magnitude[slave_index:slave_end])

        if abs(m_start-s_start)>2:
            k=1
            master_index+=m_start
            slave_index+=s_start
        else:
            k+=1
        #print(k,train_size)
        #print(master_index,slave_index)
        if master_index>=master_len or slave_index>=slave_len:
            break

        l = min(k * interval, master_len - master_index, slave_len - slave_index)

        m_tmp_end=master_index+l
        s_tmp_end=slave_index+l
        aligned_master.extend(m_magnitude[master_index:m_tmp_end])
        aligned_slave.extend(s_magnitude[slave_index:s_tmp_end])

        l=min(slave_end-slave_index,master_end-master_index)
        R=calcR(master[:,master_index:master_index+l],slave[:,slave_index:slave_index+l])
        translated_master.extend(np.dot(R,master[:,master_index:m_tmp_end]).transpose().tolist())
        translated_slave.extend((slave[:,slave_index:s_tmp_end]).transpose().tolist())


        master_index=m_tmp_end
        slave_index=s_tmp_end

    #compareGraph(master_raw,slave_raw)
    #compareGraph(translated_master,translated_slave)
    print(count*train_size/len(master_raw))
    return translated_master,translated_slave

def compareGraph(master,slave,label,f):
    plt.figure()
    plt.subplot(311)
    #print([master[i][0] for i in range(len(master))])
    labels=["device 1","device 2"]
    plt.plot([master[i][0] for i in range(len(master))],color='r')
    plt.plot([slave[i][0] for i in range(len(slave))],color='b')
    plt.ylabel(label[0])
    plt.legend(labels, fontsize=8)
    plt.subplot(312)
    plt.plot([master[i][1] for i in range(len(master))],color='r')
    plt.plot([slave[i][1] for i in range(len(slave))],color='b')
    plt.ylabel(label[1])
    plt.legend(labels, fontsize=8)
    plt.subplot(313)
    plt.plot([master[i][2] for i in range(len(master))],color='r')
    plt.plot([slave[i][2] for i in range(len(slave))],color='b')
    plt.ylabel(label[2])
    plt.xlabel("sample")
    plt.legend(labels, fontsize=8)


    plt.savefig(f)
    plt.clf()
    #plt.show()

def graphFromFile(m,s,label=["x","y","z"]):
    master=readAcc(m)
    slave=readAcc(s)
    f=m.split("\\")
    num=f[-1].split("-")[1].split(".")[0]
    f[-1]="picture"+"\\"+num+".jpg"
    figurename="\\".join(f)
    compareGraph(master,slave,label,figurename)

def test(masterfile,slavefile):
    m_a,m_g=readAccGyro(masterfile)
    s_a,s_g=readAccGyro(slavefile)

    tm,ts=adaptive_train(m_a,s_a)
    output(masterfile,tm,"AdaptiveAcc")
    output(slavefile,ts,"AdaptiveAcc")
    output(masterfile,Cartesian2Spherical(tm),"AdaptiveAccSph")
    output(slavefile,Cartesian2Spherical(ts),"AdaptiveAccSph")
    tmg,tsg=adaptive_train(m_g,s_g)
    output(masterfile,tmg,"AdaptiveGyro")
    output(slavefile,tsg,"AdaptiveGyro")
    output(masterfile,Cartesian2Spherical(tmg),"AdaptiveGyroSph")
    output(slavefile,Cartesian2Spherical(tsg),"AdaptiveGyroSph")
    return tm,ts

def output(file,data,f):
    str=file.split('\\')
    str[-2]=f
    #str[-2]="AdaptiveAcc"
    outputfile="\\".join(str)
    with open(outputfile,'w',newline="") as f:
        out=csv.writer(f)
        for i in data:
            out.writerow(i)

def main_test():
    filepath = os.getcwd()+"\\data\\transform"
    files=os.listdir(filepath)
    master="masterlocal"
    slave="slavelocal"

    for i in range(1,45):
        mfile=master+"-"+str(i)+".csv"
        sfile=slave+"-"+str(i)+".csv"
        if mfile in files and sfile in files:
            m=os.path.join(filepath,mfile)
            s=os.path.join(filepath,sfile)
            test(m,s)

def batchGraph(file):
    filepath = os.getcwd() + "\\data\\"+file
    files=os.listdir(filepath)
    master="masterlocal"
    slave="slavelocal"
    for i in range(1,45):
        mfile=master+"-"+str(i)+".csv"
        sfile=slave+"-"+str(i)+".csv"
        if mfile in files and sfile in files:
            print(mfile)
            graphFromFile(os.path.join(filepath,mfile),os.path.join(filepath,sfile))

if __name__=="__main__":
    '''m,s=test(".\\data\\transform\\masterlocal-29.csv",".\\data\\transform\\slavelocal-29.csv")
    compareGraph(m,s)
    graphFromFile(".\\data\\RMSD\\masterlocal-29.csv",".\\data\\RMSD\\slavelocal-29.csv")'''
    #test(".\\data\\transform\\masterlocal-33.csv",".\\data\\transform\\slavelocal-33.csv")
    main_test()
    batchGraph("AdaptiveAccSph")
    #batchGraph("AdaptiveAcc")
    batchGraph("AdaptiveGyroSph")
    #batchGraph("AdaptiveGyro")

File: adaptive_train.m
function adaptive_train(mfile,sfile,m_output,s_output)
    M=csvread(mfile);
    S=csvread(sfile);
    %g=M(1,16:18);
    g=M(:,16:18);
    M=M(:,1:3);
    S=S(:,1:3);
    
    m_mean=[mean(M(:,1)),mean(M(:,2)),mean(M(:,3))];
    s_mean=[mean(S(:,1)),mean(S(:,2)),mean(S(:,3))];
    M_mean=repmat(m_mean,size(M,1),1);
    S_mean=repmat(s_mean,size(S,1),1);
    M=M-M_mean;
    S=S-S_mean;
    
    norm_m=[];
    norm_s=[];
    output_M=[];
    output_S=[];
    for i=1:size(M,1)
        norm_m=[norm_m;norm(M(i,:))];
    end
    for i=1:size(S,1)
        norm_s=[norm_s;norm(S(i,:))];
    end
    m_len=size(M,1);
    s_len=size(S,1);
    %plot3D(M(1:min(m_len,s_len),:),S(1:min(m_len,s_len),:));
   %{
    figure(1)
    tl=min(m_len,s_len);
    t=1:tl;
    subplot(2,1,1);
    plot(t,M(1:tl,3),'r');hold on;
    plot(t,S(1:tl,3),'b');hold on;
    %}
    
    m_index=1;
    s_index=1;
    k=2;
    first=true;
    while true
        if first
            train_size=100;
            interval=train_size;
            first=false;
        else
             train_size=20;
             interval=50;
        end
        master_end=m_index+train_size;
        slave_end=s_index+train_size;
        
        if master_end>=m_len ||slave_end>=s_len
            %l=min(max(master_end , m_len)-min(master_end , m_len),max(slave_end , s_len)-min(slave_end , s_len));
            break
        end
        
        %[m_start,s_start]=time_align(norm_m(m_index:master_end),norm_s(s_index:slave_end));
        [m_start,s_start]=time_align(M(m_index:master_end,3),S(s_index:slave_end,3));
        if abs(m_start-s_start)>2
            k=1;
            m_index=m_index+m_start;
            s_index=s_index+s_start;
        else
            k=k+1;
        end
        
        if m_index>m_len || s_index>s_len
            break
        end
        
        l=min([k*interval,m_len-m_index,s_len-s_index]);
        
        m_tmp_end=m_index+l-1;
        s_tmp_end=s_index+l-1;

        l=min(slave_end-s_index,master_end-m_index);
        B=least_square(M(m_index:m_index+l-1,:),S(s_index:s_index+l-1,:),g(m_index));
        B
        output_M=[output_M;M(m_index:m_tmp_end,:)*B'];
        output_S=[output_S;S(s_index:s_tmp_end,:)];
        m_index=m_tmp_end;
        s_index=s_tmp_end;
        
        
        
        
        %{
        m_train_end=min(m_index+train_size-1,len);
        s_train_end=min(s_index+train_size-1,len);
        m_end=min(m_index+interval-1,len);
        s_end=min(s_index+interval-1,len);
        [m_s,s_s]=time_align(norm_m(m_index:m_train_end),norm_s(s_index:s_train_end));
        m_s=m_index+m_s;
        s_s=s_index+s_s;
        l=min(m_end-m_s,s_end-s_s);
        m_end=m_s+l;
        s_end=s_s+l;
        B=least_square(M(m_s:m_end,:),S(s_s:s_end,:));
        output_M=[output_M;M(m_s:m_end,:)*B'];
        output_S=[output_S;S(s_s:s_end,:)];
        m_index=m_end;
        s_index=s_end;
        if m_end>=len || s_end>=len ||m_train_end>=len ||s_train_end>=len
            break
        end
        %}
    end
    
    %plot3D(output_M,output_S)
    %{
    norm_ms=[];
    norm_ss=[];
    for i=1:size(output_M,1)
        norm_ms=[norm_ms;norm(output_M(i,:))];
    end
    for i=1:size(output_S,1)
        norm_ss=[norm_ss;norm(output_S(i,:))];
    end
    tl=min(size(norm_ms,1),size(norm_ss,1));
    t=1:tl;
    subplot(2,1,2);
    plot(t,norm_ms(1:tl,:),'r');hold on;
    plot(t,norm_ss(1:tl,:),'b');hold on;
    %}
    corrcoef(output_M,output_S);
    csvwrite(m_output,output_M);
    csvwrite(s_output,output_S);
end